src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc

```diff
@@ -363,6 +363,7 @@ void AresDnsResolver::OnResolvedLocked(void* arg, grpc_error* error) {
 }
 
 void AresDnsResolver::MaybeStartResolvingLocked() {
+  if (have_next_resolution_timer_) return;
```

Comment:
I don't think this is the right fix.  This will cause us to fail to immediately return results when `RequestReresolutionLocked()` is called, which we currently need to do due to deficiencies in the pick_first code (see TODO on line 388).

I think a better fix would be to change `OnResolvedLocked()` to only schedule the timer if `have_next_resolution_timer_` is false.

Reply:
Done.

Reply:
I prefer either of the following methods to fix the conflict of an incoming re-resolution and the existing timer.

1.  We can choose to keep the existing timer.
The re-resolution request will be immediately fed the cached results. But we don't start resolving until the existing timer fires.

2. We can choose to service the re-resolution request as soon as possible.
If we are not in cooldown, just cancel the existing timer and start resolving.
If we are in cooldown, cancel the existing timer and start a new timer according to the cooldown ending point. This will always make the timer fire earlier because the cooldown interval is 1s and the initial backoff interval is also 1s. 

With the current fix, we will still start resolving as long as we are not in cooldown, no matter there is a pending timer or not. But my opinion is that there is no need for the pending timer to exist if we have already start resolving. 

But I'm fine with the current fix if we need to address the issue quickly.

Reply:
I think that we never want to send a DNS request immediately if there is already a timer pending.  The whole point of the timer is to ensure that we limit the rate at which we sent DNS requests, and we lose that safety if we start choosing to issue requests immediately instead of waiting for the timer.

However, having thought about this a bit more, I think you're right that it would be better to just ignore the re-resolution request if there is already a timer pending.  So let's go ahead and do it that way (your option 1).

Reply:
Understood.